notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
-def admin_login_post():
-    if request.form['password'] != ADMIN_PASSWORD:
-        return '403 Forbidden', 403
-    session['admin'] = True
-    return redirect(external_url_for('workers'))
+        if len(pods) == 0:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod not found for given uuid")
+            return '', 200
 
+        if len(pods) > 1:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod_uuid is not unique")
 
-@app.route('/worker-image')
-def worker_image():
-    return '\n'.join(WORKER_IMAGES.values()), 200
+        for pod in pods:
+            if pod.metadata.labels['user_id'] != escp_user_id:
+                return forbidden()
+            del_pod(pod.metadata.name)
 
+        return '', 200
 
-@sockets.route('/wait')
-def wait_websocket(ws):
-    pod_name = session['pod_name']
-    svc_name = session['svc_name']
-    jupyter_token = session['jupyter_token']
-    log.info(f'received wait websocket for {svc_name} {pod_name}')
-    # wait for instance ready
-    while True:
-        try:
-            response = requests.head(f'https://notebook.hail.is/instance-ready/{svc_name}/',
-                                     timeout=1)
-            if response.status_code < 500:
-                log.info(f'HEAD on jupyter succeeded for {svc_name} {pod_name} response: {response}')
-                # if someone responds with a 2xx, 3xx, or 4xx, the notebook
-                # server is alive and functioning properly (in particular, our
-                # HEAD request will return 405 METHOD NOT ALLOWED)
-                break
-            else:
-                # somewhat unusual, means the gateway had an error before we
-                # timed out, usually means the gateway itself is broken
-                log.info(f'HEAD on jupyter failed for {svc_name} {pod_name} response: {response}')
-                gevent.sleep(1)
-            break
-        except requests.exceptions.Timeout as e:
-            log.info(f'GET on jupyter failed for {svc_name} {pod_name}')
-            gevent.sleep(1)
-    ws.send(external_url_for(f'instance/{svc_name}/?token={jupyter_token}'))
-    log.info(f'notification sent to user for {svc_name} {pod_name}')
+    except kube.client.rest.ApiException:
+        # FIXME: Return non-200 errors, not done now because web app
+        # needs to be updated to catch and handle 404 as a non-fatal response
+        # return '', str(e)[1:4]
+        return '', 200
+
+
+@app.route('/jupyter', methods=['POST'])
+def new_notebook():
+    name = request.form.get('name', 'a_notebook')
+    image = request.form.get('image', DEFAULT_HAIL_IMAGE)
+
+    user_id = request.headers.get('User')
+
+    if is_falsy_str(user_id) or image not in WORKER_IMAGES:
+        return forbidden()
+
+    # Token does not need to be cryptographically secure
+    # Is trusted, and authorization is handled upstream
```

Comment:
I think we should elide this comment because it's not a significant issue and its correctness cannot be evaluated without consulting the architecture of jupyter anyway.

Reply:
Then we should place this comment in an architectural document. It's not enough to just elide comments!